thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js

```diff
@@ -0,0 +1,303 @@
+
+    console.log('_completeRequestMissingTimeseries: merging cache');
+    return that._updateTimeseriesCache(incoming); // return new state, including new metrics
+  },
+
+  actions: {
+    addSelectedUrns(urns) {
+      console.log('addSelectedUrns()');
+      const selectedUrns = this.get('selectedUrns');
+      makeIterable(urns).forEach(urn => selectedUrns.add(urn));
+      this.set('selectedUrns', selectedUrns);
+      this.notifyPropertyChange('selectedUrns');
+    },
+
+    removeSelectedUrns(urns) {
+      console.log('removeSelectedUrns()');
+      const selectedUrns = this.get('selectedUrns');
+      makeIterable(urns).forEach(urn => selectedUrns.delete(urn));
+      this.set('selectedUrns', selectedUrns);
+      this.notifyPropertyChange('selectedUrns');
+    },
+
+    addFilteredUrns(urns) {
+      console.log('addFilteredUrns()');
+      const filteredUrns = this.get('filteredUrns');
+      makeIterable(urns).forEach(urn => filteredUrns.add(urn));
+      this.set('filteredUrns', filteredUrns);
+      this.notifyPropertyChange('filteredUrns');
+    },
+
+    removeFilteredUrns(urns) {
+      console.log('removeFilteredUrns()');
+      const filteredUrns = this.get('filteredUrns');
+      makeIterable(urns).forEach(urn => filteredUrns.delete(urn));
+      this.set('filteredUrns', filteredUrns);
+      this.notifyPropertyChange('filteredUrns');
+    },
+
+    addInvisibleUrns(urns) {
+      console.log('addInvisibleUrns()');
+      const invisibleUrns = this.get('invisibleUrns');
+      makeIterable(urns).forEach(urn => invisibleUrns.add(urn));
+      this.set('invisibleUrns', invisibleUrns);
+      this.notifyPropertyChange('invisibleUrns');
+    },
+
+    removeInvisibleUrns(urns) {
+      console.log('removeInvisibleUrns()');
+      const invisibleUrns = this.get('invisibleUrns');
+      makeIterable(urns).forEach(urn => invisibleUrns.delete(urn));
+      this.set('invisibleUrns', invisibleUrns);
+      this.notifyPropertyChange('invisibleUrns');
+    }
+  }
+});
+
+//
+// Helpers
+//
+
+const extractTimeseries = (json) => {
+  const timeseries = {};
+  Object.keys(json).forEach(range =>
+    Object.keys(json[range]).filter(sid => sid != 'timestamp').forEach(sid => {
+      const urn = `thirdeye:metric:${sid}`;
+      const jrng = json[range];
+      const jval = jrng[sid];
+
+      const timestamps = [];
+      const values = [];
+      jrng.timestamp.forEach((t, i) => {
+        if (jval[i] != null) {
+          timestamps.push(t);
+          values.push(jval[i]);
+        }
+      });
+
+      timeseries[urn] = {
+        timestamps: timestamps,
+        values: values
+      };
+    })
+  );
+  return timeseries;
+};
+
+const isIterable = (obj) => {
+  if (obj == null || _.isString(obj)) {
+    return false;
+  }
+  return typeof obj[Symbol.iterator] === 'function';
+};
+
+const makeIterable = (obj) => {
+  if (obj == null) {
+    return [];
+  }
+  return isIterable(obj) ? [...obj] : [obj];
+};
+
```

Comment:
You can extract these two helper functions into separate files in `/app/helpers` and import them here.

Reply:
Also if kept in the same file, we usually have helpers before the component

Reply:
fixed